From d3a06d091af008437e7c4a44cd38aeac83e376cd Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Wed, 18 Jun 2025 13:26:13 +0800
Subject: [PATCH 102/123] mtk: mt76: mt7996: refactor mcu commands flow for
 station mode

Fix two possible issues
1. Driver needs to update BMC STAREC before UC STAREC.
When configuring UC STAREC, it would init RA algo based on STA cap
and BSS cap. The BSS cap is stored in BMC STAREC so driver needs to
update BMC STAREC before UC STAREC.
2. Remove BMC STAREC when CHANGED_BSSID and the bssid is null.
Driver needs to remove BMC STAREC when station disconnect to the
AP. Without this patch, station cannot update the mac address when it
connect to the other AP with different mac address,

The original flow change leads to incorrect BMC WTBL settings of station
interface since some fields are not ready filled by mac80211, preventing
the firmware from parsing MU EDCA IE from the beacon and degrading MU
uplink performance.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7996/main.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/mt7996/main.c b/mt7996/main.c
index ffcf3616..d60302b9 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -979,8 +979,7 @@ mt7996_vif_cfg_changed(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 						&link->mt76, &link->msta_link,
 						true);
 			mt7996_mcu_add_sta(dev, vif, link_conf, NULL, link, NULL,
-					   CONN_STATE_PORT_SECURE,
-					   !!(changed & BSS_CHANGED_BSSID));
+					   CONN_STATE_PORT_SECURE, false);
 		}
 	}
 
@@ -1046,6 +1045,9 @@ mt7996_link_info_changed(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 				   CONN_STATE_PORT_SECURE,
 				   !!(changed & BSS_CHANGED_BSSID));
 	}
+	if (changed & BSS_CHANGED_BSSID && is_zero_ether_addr(info->bssid))
+		mt7996_mcu_add_sta(dev, vif, info, NULL, link, &link->msta_link,
+				   CONN_STATE_DISCONNECT, false);
 
 	if (changed & BSS_CHANGED_ERP_SLOT) {
 		int slottime = info->use_short_slot ? 9 : 20;
@@ -1366,6 +1368,12 @@ mt7996_mac_sta_add_links(struct mt7996_dev *dev, struct ieee80211_vif *vif,
 		if (!mconf || !conf || !link_sta)
 			continue;
 
+		/* Update bss capability after associated and before add station */
+		if (vif->type == NL80211_IFTYPE_STATION)
+			mt7996_mcu_add_bss_info(mconf->phy, vif, conf,
+						&mconf->mt76, &mconf->msta_link,
+						true);
+
 		ret = mt7996_mac_sta_init_link(dev, conf, link_sta, mconf, assoc);
 		if (ret)
 			goto error;
-- 
2.45.2

